flow/Platform.cpp

```diff
@@ -2802,39 +2802,59 @@ extern volatile void** net2backtraces;
 extern volatile size_t net2backtraces_offset;
 extern volatile size_t net2backtraces_max;
 extern volatile bool net2backtraces_overflow;
-extern volatile int net2backtraces_count;
+extern volatile int64_t net2backtraces_count;
 extern std::atomic<int64_t> net2liveness;
-extern volatile thread_local int profilingEnabled;
 extern void initProfiling();
-
-volatile thread_local bool profileThread = false;
 #endif
 
+volatile thread_local bool profileThread = false;
 volatile thread_local int profilingEnabled = 1;
 
-void setProfilingEnabled(int enabled) {
-	profilingEnabled = enabled;
+volatile thread_local int64_t numProfilesDeferred = 0;
+volatile thread_local int64_t numProfilesOverflowed = 0;
+volatile thread_local int64_t numProfilesCaptured = 0;
+volatile thread_local bool profileRequested = false;
+
+int64_t getNumProfilesDeferred() {
+	return numProfilesDeferred;
+}
+
+int64_t getNumProfilesOverflowed() {
+	return numProfilesOverflowed;
+}
+
+int64_t getNumProfilesCaptured() {
+	return numProfilesCaptured;
 }
 
 void profileHandler(int sig) {
 #ifdef __linux__
-	if (!profileThread || !profilingEnabled) {
+	if(!profileThread) { 
+		return;
+	}
+
+	if(!profilingEnabled) {
+		profileRequested = true;
+		++numProfilesDeferred;
 		return;
 	}
 
-	net2backtraces_count++;
+	++net2backtraces_count;
+
 	if (!net2backtraces || net2backtraces_max - net2backtraces_offset < 50) {
+		++numProfilesOverflowed;
 		net2backtraces_overflow = true;
 		return;
 	}
 
+	++numProfilesCaptured;
+
 	// We are casting away the volatile-ness of the backtrace array, but we believe that should be reasonably safe in the signal handler
 	ProfilingSample* ps = const_cast<ProfilingSample*>((volatile ProfilingSample*)(net2backtraces + net2backtraces_offset));
 
 	ps->timestamp = timer();
 
```

Comment:
The description in your PR made it sound like deferred profiles with then later be collected and attributed somewhere, but I don't see us recording this sample N times and resetting numProfilesDeferred to 0 ?

I'd think we'd either skip collecting them all together, and accept that flow profiling will have a blindspot for resolving dynamically linked library things, or to attribute the stacks to whatever then flips profiling to on for the thread, which should be whatever requested the expensive dynamically linked library thing.  Doing this registers one stacktrace for dl_iterate_phdr, and numProfilesDeferred would show the inaccuracy, but it means an extra thing to pay attention to.

I'd sort of be more in favor of fully going one way or the other, though I don't think that this is really an important thing that needs to be resolved right now.  I'd be fine merging this now to get performance/correctness results and change stack collection later.

Reply:
`numProfilesDeferred` (and other counters) are just running totals that I think are used only for logging during testing. `profileRequested` tracks whether a profile was attempted while disabled, but you make an interesting point about tracking the number of samples requested.

You can see the code in `setProfilingEnabled` that registers a stacktrace when turning profiling back on if one was attempted while it was disabled.

Reply:
Yes, but it only registers *a* trace, regardless of how many were requested while dl_iterate_phdr was running.  I'm inclined to say we either shouldn't register a trace, or should register N traces for N requests, but registering 1 trace for N requests feels a bit odd to me.

Reply:
I again, however, do not feel strongly.  So I'll go ahead and merge now anyway, and we can talk about if this matters enough to actually address it later.

Reply:
Yeah, understood, I hadn't thought about that initially but think it's a good point. I think there's value in having 1 rather than 0 for slow task profiling, but done right I think N would be better.